### PR TITLE
[runtime] New profiler API.

### PR DESCRIPTION
**Pull request by alexrp**
This is my proposed replacement for the old profiler API. Sorry for the huge commit, but with the far-reaching changes, it was difficult to split the changes up in actually atomic commits. I recommend viewing the `mono/metadata/profiler*` files normally rather than looking at the diff as they've been almost 100% rewritten.

# TL;DR

Old code:

```c
void
mono_profiler_startup (const char *desc)
{
    int events = MONO_PROFILE_NONE;

    MonoProfiler *prof = calloc (1, sizeof (MonoProfiler));
    mono_profiler_install (prof, shutdown_cb);

    events |= MONO_PROFILE_THREADS;
    mono_profiler_install_thread (thread_start_cb, thread_end_cb);

    events |= MONO_PROFILE_GC;
    mono_profiler_install_gc (gc_event_cb, NULL);

    events |= MONO_PROFILE_ALLOCATIONS;
    mono_profiler_install_allocation (gc_alloc_cb);

    events |= MONO_PROFILE_STATISTICAL;
    mono_profiler_install_statistical (stat_hit_cb);
    mono_profiler_set_statistical_mode (MONO_PROFILER_STAT_MODE_PROCESS, 100);

    mono_profiler_set_events (events);
}
```

New code:

```c
void
mono_profiler_init (const char *desc)
{
    MonoProfiler *prof = calloc (1, sizeof (MonoProfiler));
    MonoProfilerHandle handle = mono_profiler_install (prof);

    mono_profiler_set_runtime_shutdown_callback (handle, shutdown_cb);
    mono_profiler_set_thread_started_callback (handle, thread_start_cb);
    mono_profiler_set_thread_stopped_callback (handle, thread_end_cb);
    mono_profiler_set_gc_event_callback (handle, gc_event_cb);
    mono_profiler_set_gc_allocation_callback (handle, gc_alloc_cb);
    mono_profiler_set_sample_hit_callback (handle, stat_hit_cb);

    mono_profiler_enable_allocations ();

    if (mono_profiler_enable_sampling (handle))
        mono_profiler_set_sample_mode (handle, MONO_PROFILER_SAMPLE_MODE_PROCESS, 100);
    else
        g_warning ("Another profiler controls sampling parameters");
}
```

# The Long Story

`mono_profiler_install` now returns a `MonoProfilerHandle` which, to users, is an opaque pointer. This is passed to other API functions to set callbacks and change various settings (e.g. sampling parameters). This allows a profiler to change its setup long after runtime initialization.

Additionally, some profiling features now have to be enabled explicitly in a profiler's `mono_profiler_init` function (referred to as the init function). Sampling, for example, must be enabled by calling `mono_profiler_enable_sampling` in the init function. Whichever profiler calls this function first gets exclusive control over sampling parameters (mode and frequency) through `mono_profiler_set_sample_mode`. Other profilers can still receive sample hit callbacks, they just can't change the mode or frequency. Simply calling `mono_profiler_enable_sampling` does not make the sampling thread do anything; it just puts it in an idle mode for later use when the controlling profiler sets the sample mode to something other than `MONO_PROFILER_SAMPLE_MODE_NONE`. Allocations also must be explicitly enabled with `mono_profiler_enable_allocations`. This is mainly so that the runtime knows to instrument managed allocators with a call to the profiler API at the end.

The coverage API has not changed much beyond a few tweaks for naming consistency, however, the old coverage mode that was implemented on a per-arch basis has been completely removed in favor of the IR-based mode.

Enter/leave instrumentation is now more fine-grained. Like coverage, a filter callback can be installed which can decide whether to instrument a given method's prologue, epilogue, or both.

The API doesn't currently have a way to enable 'heavy' features (sampling, allocations, enter/leave) after runtime initialization. For sampling, this is a non-issue as you should just be enabling sampling in the idle mode if you plan to use it later. The most it'll cost you is a thread that's sleeping on a semaphore. For allocations, we can always change the API contract of `mono_profiler_enable_allocations` to allow it to be called at any point, where it could then recompile managed allocators or something. Similarly, for enter/leave instrumentation, a `mono_profiler_recompile_world` function could be introduced if/when we engineer such a feature into the JIT (and if we decide that it's actually a good idea to have such a feature in the profiler API).

# The Implementation

The vast majority of the profiler callback API is now generated from macros in a single file, `mono/metadata/profiler-events.h`. This makes it super trivial to add new callbacks. Literally all you need to do is add a line in that file and invoke `MONO_PROFILER_RAISE` wherever you want to raise the event. The one aspect of this that I'm not so sure about is generating the public API in `mono/metadata/profiler.h` based on these macros. It's super convenient, but it does mean a user of the API has to read a little more closely to understand how the functions there are generated. I have provided a few examples in a comment to give the general idea, though. Either way, we could declare stuff there manually if people are in favor of doing that instead.

The API no longer uses flags to enable events. Instead, a sort of reference counter (see `update_callback` in `mono/metadata/profiler.c`) is maintained for each kind of callback. This allows us to do the same thing we did with `mono_profiler_get_events` in the past, i.e. not doing expensive data gathering or calling into `mono_profiler_raise_*` functions if no callbacks are registered.

If a profiler module with the old entry point name (`mono_profiler_startup`, as opposed to the new `mono_profiler_init`) is loaded, a warning will be printed and the runtime will refuse to call into it. This way, developers will easily recognize if their module hasn't been migrated to the new API.

As I mentioned above, managed allocators are now instrumented when allocation profiling is enabled. This means that we no longer use the slow path variant that simply calls directly into the runtime to use the C allocation functions. Instead, we just do a call to the profiler API at the end of the managed allocator. This is a huge performance improvement and will especially affect people who do sampling and allocation profiling at the same time (i.e. ~everybody).

### CODE REVIEW

**Comment by alexrp** [comment]
I've added f23ecd6276b8d1da75dedc7a410078422c9e6af7 as an example of how to add new events with this API.

File: mono/metadata/boehm-gc.c [hunk 1]
```diff
@@ -1129,7 +1130,7 @@ mono_gc_get_managed_allocator (MonoClass *klass, gboolean for_box, gboolean know
 		return NULL;
 	if (mono_class_has_finalizer (klass) || mono_class_is_marshalbyref (klass))
 		return NULL;
-	if (mono_profiler_get_events () & (MONO_PROFILE_ALLOCATIONS | MONO_PROFILE_STATISTICAL))
+	if (G_UNLIKELY (mono_profiler_allocations_enabled ()))
```
Comment (alexrp) [inline mono/metadata/boehm-gc.c]: I only implemented it for SGen. Boehm will just continue to use the native allocation functions as it always has. Well, unless there's some compelling reason why we should bother with adding new features to Boehm.

File: mono/metadata/boehm-gc.c [hunk 2]
```diff
@@ -1680,7 +1681,7 @@ alloc_handle (HandleData *handles, MonoObject *obj, gboolean track)
 #endif
 	unlock_handles (handles);
 	res = MONO_GC_HANDLE (slot, handles->type);
-	mono_profiler_gc_handle (MONO_PROFILER_GC_HANDLE_CREATED, handles->type, res, obj);
+	MONO_PROFILER_RAISE (gc_handle_created, (res, handles->type, obj));
```
Comment (alexrp) [inline mono/metadata/boehm-gc.c]: Intended. I thought this argument ordering made a lot more sense.

File: mono/metadata/loader.c [hunk 3]
```diff
@@ -1901,11 +1901,10 @@ mono_get_method_constrained_checked (MonoImage *image, guint32 token, MonoClass
 void
 mono_free_method  (MonoMethod *method)
 {
-	if (mono_profiler_get_events () & MONO_PROFILE_METHOD_EVENTS)
-		mono_profiler_method_free (method);
+	MONO_PROFILER_RAISE (method_free, (method));
 	
 	/* FIXME: This hack will go away when the profiler will support freeing methods */
-	if (mono_profiler_get_events () != MONO_PROFILE_NONE)
+	if (G_UNLIKELY (mono_profiler_installed ()))
```
Comment (alexrp) [inline mono/metadata/loader.c]: Yes, due to the method JIT event hack in the profiler, which probably isn't going away anytime soon.

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -7304,8 +7304,7 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 	cfg->dont_inline = g_list_prepend (cfg->dont_inline, method);
 	if (cfg->method == method) {
 
-		if (cfg->prof_options & MONO_PROFILE_INS_COVERAGE)
-			cfg->coverage_info = mono_profiler_coverage_alloc (cfg->method, header->code_size);
+		cfg->coverage_info = mono_profiler_coverage_alloc (cfg->method, header->code_size);
```
Comment (alexrp) [inline mono/mini/method-to-ir.c]: The installed filter functions will determine if anything actually gets allocated here. If no filter functions are allocated, `mono_profiler_coverage_alloc` returns `NULL`.

File: mono/mini/interp/interp.c [hunk 5]
```diff
@@ -173,8 +173,8 @@ debug_enter (MonoInvocation *frame, int *tracing)
 		g_print  ("%s)\n", args);
 		g_free (args);
 	}
-	if (mono_profiler_events & MONO_PROFILE_ENTER_LEAVE)
-		mono_profiler_method_enter (frame->runtime_method->method);
+	if (mono_profiler_should_instrument_method (frame->runtime_method->method, TRUE))
```
Comment (alexrp) [inline mono/mini/interp/interp.c]: That's implicit. If no filter functions are installed, enter/leave is not enabled. I don't think requiring it to be explicitly enabled adds much value compared to just installing a filter function if you want to use the feature.

File: mono/metadata/sgen-mono.c [hunk 6]
```diff
@@ -1816,8 +1840,8 @@ mono_gc_alloc_string (MonoVTable *vtable, size_t size, gint32 len)
 	UNLOCK_GC;
 
  done:
-	if (G_UNLIKELY (mono_profiler_events & MONO_PROFILE_ALLOCATIONS))
-		mono_profiler_allocation (&str->object);
+	if (G_UNLIKELY (mono_profiler_allocations_enabled ()))
```
Comment (alexrp) [inline mono/metadata/sgen-mono.c]: I think that would be confusing behavior. In that case, if a profiler installs an allocation callback but doesn't call `mono_profiler_enable_allocations`, it'd be notified of allocations from C code, but not from those done in managed allocators. It seems better to require a profiler to call `mono_profiler_enable_allocations` if it wants allocation notifications *at all*.

File: mono/mini/interp/interp.c (hunk 5 above)
Comment (alexrp) [inline mono/mini/interp/interp.c]: Does that really add any value? "Enabling" enter/leave, in the new API, means installing a filter function that gets called here. If no filter function is installed, no enter/leave instrumentation is done.

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (alexrp) [inline mono/mini/method-to-ir.c]: Similar to enter/leave, `mono_profiler_coverage_alloc` will return `NULL` if no filters are installed or if none of them request the given method to be instrumented.

File: mono/metadata/boehm-gc.c (hunk 1 above)
Comment (kumpera) [inline mono/metadata/boehm-gc.c]: Oh! This is boehm. 🤦‍♀️

File: mono/metadata/loader.c (hunk 3 above)
Comment (kumpera) [inline mono/metadata/loader.c]: Right, is it something worth fixing?

File: mono/metadata/sgen-mono.c (hunk 6 above)
Comment (kumpera) [inline mono/metadata/sgen-mono.c]: You're right, a single cond would break how the API is supposed to be used.

File: mono/metadata/loader.c (hunk 3 above)
Comment (alexrp) [inline mono/metadata/loader.c]: I *think* this hack only affects `DynamicMethod`s, but not 100% sure. If that's the case, I would say it's not worth spending time on right now. I'll file a bug for it in any case, though, so it's not forgotten.

**Comment by knocte** [comment]
@alexrp hey sorry for the slight off-topic, but do you know if these days .NET6 (or upcoming version) has a profiler API similar to this?

What I would want to achieve is something similar to this: https://github.com/slluis/gui-thread-check , but for .NET instead of Mono.

Thanks

**Comment by alexrp** [comment]
@knocte CoreCLR has a profiler API but it's very old and based on nano-COM, so it's not exactly fun to work with.

Some useful links:

* https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/
* https://github.com/dotnet/runtime/tree/main/docs/design/coreclr/profiling
* https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/profiling.md
* https://github.com/microsoft/clr-samples/tree/master/ProfilingAPI

There is also the newer [EventPipe](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/eventpipe) but it is not quite as powerful as the full profiler API (e.g. no enter/leave callbacks). It's more akin to Mono's log profiler.

**Pull request merged**